jlm/llvm/opt/alias-analyses/LocalAliasAnalysis.hpp

```diff
@@ -0,0 +1,266 @@
+   * @param N the number of bytes into its memory region an access must be to be kept.
+   */
+  static void
+  RemoveTopOriginsWithinTheFirstNBytes(TraceCollection & traces, size_t s, size_t N);
+
+  /**
+   * Checks if any of the top origins in the two trace collections are the same,
+   * and have overlapping offsets.
+   * Only identical top origins are considered, so if two distinct top origins
+   * point to the same memory, that aliasing will not be detected by this method.
+   *
+   * @param tc1 the collection of traces of p1
+   * @param s1 the size of the operation performed at p1
+   * @param tc2 the collection of traces of p2
+   * @param s2 the size of the operation performed at p2
+   * @return true if any top origins are shared, and possibly overlap
+   */
+  [[nodiscard]] static bool
+  DoTraceCollectionsOverlap(TraceCollection & tc1, size_t s1, TraceCollection & tc2, size_t s2);
+
+  /**
+   * Checks if the given pointer is the output of an original memory location,
+   * AND that the address of the memory location is never passed anywhere that is not
+   * traceable back to the original operation.
+   *
+   * Only \ref AllocaOperation is fully traceable, and only when the address can be traced to all
+   * uses, and all uses are loads and stores. If the address is passed to a function, or stored in
+   * a variable, the \ref AllocaOperation is not fully traceable.
+   * For a fully traceable \ref AllocaOperation, any use of its address can therefore be traced back
+   * to the \ref AllocaOperation using the \ref TraceAllPointerOrigins function
+   *
+   * In summary: this function performs a simple, local, escape analysis for ALLOCAs.
+   * Its result is cached for performance.
+   *
+   * @param pointer the pointer output to be analyzed
+   * @return true if the output is an original pointer, and it can be fully traced
+   */
+  [[nodiscard]] bool
+  IsOriginalOriginFullyTraceable(const rvsdg::Output & pointer);
+
+  /**
+   * Checks if the given trace collection only contains top origins that are fully traced.
+   * @param traces the trace collection
+   * @return true if all top origins are fully traceable
+   */
+  [[nodiscard]] bool
+  HasOnlyFullyTraceableTopOrigins(TraceCollection & traces);
+
+  /**
+   * Memoization of "fully traceable" (escape analysis) queries.
+   * It assumes that no changes are made to the underlying RVSDG between queries.
+   */
+  std::unordered_map<const rvsdg::Output *, bool> IsFullyTraceable_;
+};
+
+/**
+ * Gets the value of the given \p output as a compile time constant, if possible.
+ * The constant is interpreted as a signed value, and sign extended to int64 if needed.
+ * This function does not perform any constant folding.
+ *
+ * @param output the output whose constant value is requested
+ * @return the value of the output, or nullopt if it could not be determined.
+ */
+[[nodiscard]] std::optional<int64_t>
+GetConstantSignedIntegerValue(const rvsdg::Output & output);
```

Comment:
Does it need to be exposed? If it is private to LocalAliasAnalysis, then maybe not expose in header. Other than that, the concept sounds "fairly generic" and not related to alias analysis at all to me.